Freezing pre-connection state. (#860)
ietf-tapswg/api-drafts: draft-ietf-taps-impl.md

- the Connection, the basic object that represents a flow of data as Messages in either direction between the Local and Remote Endpoints;
- and the Listener, a passive waiting object that delivers new Connections.

Preconnection objects should be implemented as bundles of properties that an application can both read and write. An implementation could allow applications to update the properties or endpoints of a preconnection. However, an application should avoid changes to the Preconnection properties and endpoints to improve consistent operation across different implementations.

Connection objects represent the interface between the application and the implementation to manage transport state, and conduct data transfer. During the process of establishment ({{conn-establish}}), the Connection will not be bound to a specific transport flow, since multiple candidate Protocol Stacks might be raced. 

> Once the Connection is established, its interface maps actions and events to the details of the chosen Protocol Stack. For example, the same Connection object may ultimately interface to a single instance of one transport protocol (e.g., a TCP connection, a TLS session over TCP, a UDP flow with fully-specified Local and Remote Endpoints, a DTLS session, a SCTP stream, a QUIC stream, or an HTTP/2 stream).

Reviewer: Here, the original text said "may ultimately represent the interface into", and the proposed text now says "may ultimately interface to"...  all this reads a bit clumsy to me.

Could we not just say "may ultimately represent a single instance ......"
Maybe it's a bit vague, that's in the nature of the word "represent", but I think the meaning is clear enough and this would read better.

Author: Could is /use/ better than /represent/?:  
(1) For example, the same Connection object may ultimately use a single instance of one transport protocol

> The implementation should ensure that the copy of all Properties held by a Connection or Listener is independent of other connections. This may involve performing a deep-copy, copying the object with all the objects it references. Once Initate has been called, the Selection Properties and Endpoint information are immutable (i.e, an application is not able to later modify Selection Properties on the original Preconnection object).

Reviewer: I see that this adds the explicit parts from the original text. I like it for being clear about what's mutable and what isn't, and I still think that the whole thing would be best with my suggested text change above - this here just adds further clarification.

The first sentence of this paragraph says "... is independent of other connections."  I guess this should really say ""... is independent of any other connections that are not part of the same Connection Group."

Author: I think this was done.

- the Connection, the basic object that represents a flow of data as Messages in either direction between the Local and Remote Endpoints;
- and the Listener, a passive waiting object that delivers new Connections.

> Preconnection objects should be implemented as bundles of properties that an application can both read and write. An implementation could allow properties of a preconnection to be updated. However, an application should avoid changes to the Preconnection Selection and Connection to improve consistent operation across different implementations.

Reviewer 2: ```suggestion
Preconnection objects should be implemented as bundles of properties that an application can both read and write. An implementation could allow applications to update the properties or endpoints of a preconnection after it is initiated. However, an application should avoid changes to the Preconnection properties and endpoints to improve consistent operation across different implementations.
```

> 

Reviewer 3: The original seems clearer to me: it's more explicit about what happens, and when.

Author: Oh dear we don't seem to agree - I think the original hints at everything being immutable. Instead, I think everything should be copied; but only the selection params are immutable, the other params are mutable on the copy.

Reviewer 3: That would be fine, but I didn't get that meaning from the new text.

Author: So... what's missing? - The text on copying was moved to the section after "once a connection is established.

Reviewer 3: I can't parse "An implementation could allow properties of a preconnection to be updated, but changes to the Preconnection Selection and Connection may be language dependent and is to be avoided in code that could be used across different implementations". I read it as trying to simplify the original wording, but actually making it less clear to me. I don't see how it relates to the selection parameters being immutable and while the other parameters remain mutable.

Author: Improved text

Reviewer 3: The last two sentences here still seem vague and contradictory with the below. I think it'd be clearer just to remove them, and leave the detailed following explanations.

Author: @martinduke :  The expected operation is now clearer, csp suggests simply not mentioning the "update during pre-connection" possibility. What do you think?

Reviewer: I have to agree with @csperkins that the original was clearer, mainly for being more explicit - but I think that, after looking at this for some time, I now see where the problem is (which @csperkins is perhaps missing, and which @gorryfair is perhaps trying to fix).

I suspect that @gorryfair wants to avoid that we say, here, that properties on a Connection cannot be changed after creating an outbound Connection or a Listener. This would, of course, be wrong, but the text says that they are all "immutable", which gives a wrong indication.

Then again, the proposed text now doesn't look like a good fix to me. E.g., it says "an implementation could allow applications to...", followed by "however, applications should not do this". This is awkward. Maybe you meant to write "Connection", not "preconnection" in the  second sentence?

Here's a proposal: a change to the original text that avoids falsely indicating that everything becomes immutable:

***
Preconnection objects should be implemented as bundles of properties that an application can both read and write. A Preconnection object influences a Connection only at one point in time: when the Connection is created. Once a Preconnection has been used to create an outbound Connection or a Listener, the implementation should ensure that the copy of the properties held by the Connection or Listener is not affected when the application makes changes to the Preconnection object. This may involve performing a deep-copy, copying the object with all the objects it references.
***

I also removed the trailing "if the application is still able..." because I don't think this added clarity.

Author: I tried, and merged. Please check the new diff and suggest improvements.

Reviewer: FWIW, I agree with this update. Thanks!
